optimizer: simplify inlining algorithm

aviatesk opened the pull request:
The main purpose of this commit is to try to simplify our inlining
algorithm by removing the delayed callee resolving mechanism (i.e.
`DelayedInliningSpec` stuff), since that kind of mechanism currently
has no users and we provide more fine grained control with
`inlining_policy` now.

This commit also adds some more cleanups, changing argument ordering of
some subroutines so that they are aligned with the other ones, and
removing dead subroutines.

nanosoldier commented:
[Your benchmark job](https://github.com/JuliaLang/julia/pull/47207) has completed - possible performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/NanosoldierReports/blob/master/benchmark/by_hash/003bea5_vs_dc3a2e8/report.md).

aviatesk committed:
optimizer: simplify inlining algorithm

The main purpose of this commit is to try to simplify our inlining
algorithm by removing the delayed callee resolving mechanism (i.e.
`DelayedInliningSpec` stuff), since that kind of mechanism currently
has no users and we provide more fine grained control with
`inlining_policy` now.

This commit also adds some more cleanups, changing argument ordering of
some subroutines so that they are aligned with the other ones, and
removing dead subroutines.

aviatesk commented:
@nanosoldier `runbenchmarks("inference", vs=":master")`

nanosoldier commented:
[Your benchmark job](https://github.com/JuliaLang/julia/pull/47207#issuecomment-1282042391) has completed - possible performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/NanosoldierReports/blob/master/benchmark/by_hash/26b5b4a_vs_dc3a2e8/report.md).

aviatesk commented:
Okay, good performance. This should be ready.

aviatesk commented:
@nanosoldier `runbenchmarks("inference", vs=":master")`

nanosoldier commented:
[Your benchmark job](https://github.com/JuliaLang/julia/pull/47207#issuecomment-1282150431) has completed - possible performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/NanosoldierReports/blob/master/benchmark/by_hash/6bf54c7_vs_b55c15e/report.md).

aviatesk commented:
@nanosoldier `runbenchmarks("inference", vs=":master")`

nanosoldier commented:
[Your benchmark job](https://github.com/JuliaLang/julia/pull/47207#issuecomment-1283018011) has completed - no performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/NanosoldierReports/blob/master/benchmark/by_hash/35b2f16_vs_61ccb32/report.md).
